File: tests/unit/ast/node.cpp
//   Copyright (c) 2017 Hartmut Kaiser
//
//   Distributed under the Boost Software License, Version 1.0. (See accompanying
//   file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <phylanx/phylanx.hpp>

#include <hpx/hpx_main.hpp>
#include <hpx/testing.hpp>

#include <cstdint>
#include <utility>
#include <vector>

#include <blaze/Math.h>

///////////////////////////////////////////////////////////////////////////////
template <typename Ast>
void test_serialization(Ast const& in)
{
    std::vector<char> buffer = phylanx::util::serialize(in);
    Ast out = phylanx::util::unserialize<Ast>(buffer);

    HPX_TEST_EQ(in, out);
}

///////////////////////////////////////////////////////////////////////////////
void test_identifier()
{
    phylanx::ast::identifier id("some_name");
    test_serialization(id);
}

///////////////////////////////////////////////////////////////////////////////
void test_primary_expr()
{
    phylanx::ast::primary_expr p1(true);
    test_serialization(p1);

    phylanx::ast::primary_expr p2(
        phylanx::ast::identifier("some_name"));
    test_serialization(p2);

    blaze::Rand<blaze::DynamicVector<double>> gen{};
    blaze::DynamicVector<double> v = gen.generate(1007UL);
    phylanx::ast::primary_expr p3{
        phylanx::ir::node_data<double>{v}};
    test_serialization(p3);

    phylanx::ast::primary_expr p4{"some string"};
    test_serialization(p4);

    phylanx::ast::primary_expr p5{std::uint64_t(42)};
    test_serialization(p5);
}

///////////////////////////////////////////////////////////////////////////////
void test_operand()
{
    phylanx::ast::primary_expr p1(true);
    phylanx::ast::operand op1(p1);
    test_serialization(op1);

    phylanx::ast::primary_expr p2(
        phylanx::ast::identifier{"some_name"});
    phylanx::ast::operand op2(std::move(p2));
    test_serialization(op2);

    phylanx::ast::primary_expr p3{"some string"};
    phylanx::ast::operand op3(std::move(p3));
    test_serialization(op3);

    phylanx::ast::primary_expr p4{std::uint64_t(42)};
    phylanx::ast::operand op4(std::move(p4));
    test_serialization(op4);
}

///////////////////////////////////////////////////////////////////////////////
void test_unary_expr()
{
    phylanx::ast::primary_expr p1(true);
    phylanx::ast::operand op1(p1);
    phylanx::ast::unary_expr u1(phylanx::ast::optoken::op_not, op1);
    test_serialization(u1);

    phylanx::ast::primary_expr p2(
        phylanx::ast::identifier{"some_name"});
    phylanx::ast::operand op2(p2);
    phylanx::ast::unary_expr u2(
        phylanx::ast::optoken::op_negative, std::move(op2));
    test_serialization(u2);
}

///////////////////////////////////////////////////////////////////////////////
void test_operation()
{
    phylanx::ast::primary_expr p1(true);
    phylanx::ast::operand op1(p1);
    phylanx::ast::operation u1(phylanx::ast::optoken::op_plus, op1);
    test_serialization(u1);

    phylanx::ast::primary_expr p2(
        phylanx::ast::identifier{"some_name"});
    phylanx::ast::operand op2(p2);
    phylanx::ast::operation u2(
        phylanx::ast::optoken::op_minus, std::move(op2));
    test_serialization(u2);
}

///////////////////////////////////////////////////////////////////////////////
void test_expression()
{
    phylanx::ast::primary_expr p1(true);
    phylanx::ast::operand op1(p1);
    phylanx::ast::expression e1(op1);

    phylanx::ast::operation u1(phylanx::ast::optoken::op_plus, std::move(op1));
    e1.append(u1);

    std::vector<phylanx::ast::operation> ops = {u1, u1};
    e1.append(ops);

    test_serialization(e1);

    phylanx::ast::primary_expr p2(std::move(e1));
    test_serialization(p2);
}

void test_function_call()
{
    phylanx::ast::identifier id("function_name");
    phylanx::ast::function_call fc(std::move(id));

    phylanx::ast::primary_expr p1(true);
    phylanx::ast::operand op1(p1);
    phylanx::ast::expression e1(std::move(p1));

    fc.append(e1);

    std::vector<phylanx::ast::expression> exprs = {e1, e1};
    fc.append(exprs);

    test_serialization(fc);

    phylanx::ast::primary_expr p2(std::move(fc));
    test_serialization(p2);
}

void test_lists()
{
    phylanx::ast::identifier id("function_name");
    phylanx::ast::function_call fc(std::move(id));

    phylanx::ast::primary_expr p1(true);
    phylanx::ast::operand op1(p1);
    phylanx::ast::expression e1(std::move(p1));
    fc.append(e1);

    std::vector<phylanx::ast::expression> exprs = {e1, e1};
    fc.append(exprs);

    std::vector<phylanx::ast::expression> list = {
        phylanx::ast::expression{fc},
        phylanx::ast::expression{op1}
    };

    test_serialization(list);
}

///////////////////////////////////////////////////////////////////////////////
int main(int argc, char* argv[])
{
    test_identifier();
    test_primary_expr();
    test_operand();
    test_unary_expr();
    test_operation();
    test_expression();
    test_function_call();
    test_lists();

    return hpx::util::report_errors();
}

File: tests/unit/python/ast/node.py
#  Copyright (c) 2017 Hartmut Kaiser
#
#  Distributed under the Boost Software License, Version 1.0. (See accompanying
#  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

import phylanx
import phylanx.ast as ast
import phylanx.util as util

###############################################################################

phylanx.PhylanxSession.init(1)


def test_serialization(in_ast):
    data = util.serialize(in_ast)
    out_ast = util.unserialize_expr(data)
    assert (in_ast == out_ast)

    import pickle
    pickled_data = pickle.dumps(in_ast, pickle.DEFAULT_PROTOCOL)
    out_ast = pickle.loads(pickled_data)
    assert (in_ast == out_ast)


###############################################################################


def test_identifier():
    ident1 = ast.identifier('ident1')
    assert (ident1.name == 'ident1')

    ident2 = ast.identifier('ident1')
    assert (ident2.name == 'ident1')
    assert (ident1 == ident2)

    ident3 = ast.identifier('ident3')
    assert (ident3.name == 'ident3')
    assert (ident1 != ident3)


def test_primary_expr():
    pe1 = ast.primary_expr(False)
    assert (not pe1.value)

    ident1 = ast.identifier('ident1')
    pe2 = ast.primary_expr(ident1)
    assert (pe2.value == ident1)


def test_operand():
    pe1 = ast.primary_expr(True)
    op1 = ast.operand(pe1)
    assert (op1.value == pe1)
    assert (op1.value.value)

    id2 = ast.identifier('ident')
    pe2 = ast.primary_expr(id2)
    op2 = ast.operand(pe2)
    assert (op2.value == pe2)
    assert (op2.value.value.name == 'ident')


def test_unary_expr():
    pe1 = ast.primary_expr(True)
    op1 = ast.operand(pe1)
    ue1 = ast.unary_expr(ast.optoken.op_not, op1)
    assert (ue1.operator == ast.optoken.op_not)
    assert (ue1.operand == op1)

    id2 = ast.identifier('ident')
    pe2 = ast.primary_expr(id2)
    op2 = ast.operand(pe2)
    ue2 = ast.unary_expr(ast.optoken.op_negative, op2)
    assert (ue2.operator == ast.optoken.op_negative)
    assert (ue2.operand == op2)


def test_operation():
    pe1 = ast.primary_expr(True)
    op1 = ast.operand(pe1)
    op2 = ast.operation(ast.optoken.op_plus, op1)
    assert (op2.operator == ast.optoken.op_plus)
    assert (op2.operand == op1)

    id2 = ast.identifier('ident')
    pe2 = ast.primary_expr(id2)
    op3 = ast.operand(pe2)
    op4 = ast.unary_expr(ast.optoken.op_minus, op3)
    assert (op4.operator == ast.optoken.op_minus)
    assert (op4.operand == op3)


def test_expression():
    pe1 = ast.primary_expr(True)
    op1 = ast.operand(pe1)
    e1 = ast.expression(op1)

    op2 = ast.operation(ast.optoken.op_plus, op1)
    e1.rest.append(op2)
    test_serialization(e1)

    e1.rest.append(op2)
    e1.rest.append(op2)
    test_serialization(e1)


###############################################################################
test_identifier()
test_primary_expr()
test_operand()
test_unary_expr()
test_operation()
test_expression()
